vault/resource_secrets_sync_association.go

```diff
@@ -0,0 +1,177 @@
+		ReadContext:   provider.ReadContextWrapper(secretsSyncAssociationRead),
+		DeleteContext: secretsSyncAssociationDelete,
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of the store/destination.",
+				ForceNew:    true,
+			},
+			consts.FieldType: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Type of secrets store.",
+				ForceNew:    true,
+			},
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the mount where the secret is located.",
+			},
+			fieldAccessor: {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "Specifies the unique accessor of the mount.",
+				ValidateFunc: provider.ValidateNoLeadingTrailingSlashes,
+			},
+			fieldSecretName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the name of the secret to synchronize.",
+			},
+			fieldSyncStatus: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Specifies the status of the association.",
+			},
+			fieldUpdatedAt: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Duration string stating when the secret was last updated.",
+			},
+		},
+	}
+}
+
+func secretsSyncAssociationWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	name := d.Get(consts.FieldName).(string)
+	destType := d.Get(consts.FieldType).(string)
+	path := secretsSyncAssociationSetPath(name, destType)
+
+	data := map[string]interface{}{}
+
+	for _, k := range []string{
+		fieldSecretName,
+		consts.FieldMount,
+	} {
+		data[k] = d.Get(k)
+	}
+
+	log.Printf("[DEBUG] Writing association to %q", path)
+	resp, err := client.Logical().Write(path, data)
+	if err != nil {
+		return diag.Errorf("error setting secrets sync association %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Wrote association to %q", path)
+
+	// expect accessor to be provided from mount
+	accessor := d.Get(fieldAccessor).(string)
+
+	// associations can't be read from Vault
+	// set data that is received from Vault upon writes
+	vaultRespKey := fmt.Sprintf("%s/%s", accessor, name)
+	associatedSecrets := "associated_secrets"
+	for _, k := range secretsSyncAssociationFields {
+		if secrets, ok := resp.Data[associatedSecrets]; ok {
+			if secretsMap, ok := secrets.(map[string]interface{}); ok {
+				if val, ok := secretsMap[vaultRespKey]; ok {
+					if valMap, ok := val.(map[string]interface{}); ok {
+						if v, ok := valMap[k]; ok {
+							if err := d.Set(k, v); err != nil {
+								return diag.FromErr(err)
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
Thoughts on rewriting this to fail fast instead of nesting?
eg:
```go
secrets, ok := resp.Data[associatedSecrets]
if !ok {
    continue
}
...
```

Reply:
Good call, updated to break early with an error if no associated secret is found even though it is expected